### PR TITLE
Add UI for switching public mail notification language

### PR DESCRIPTION
**Pull request by VicDeo**
## Description
- [x] Make lang selector reusable
- [x] Reuse it for a new option in the `Sharing` section

## Related Issue
https://github.com/owncloud/enterprise/issues/2435#issuecomment-396551153

## Motivation and Context
Allows admin users change a language used for public mail notifications

## How Has This Been Tested?
0. Language selector in Settings->Personal -> General -> Language should NOT change it's behavior
1. A new option should be visible only when `Allow users to send mail notification for shared files` is checked in Settings->Admin->Sharing
2. Switching the dropdown  to  the value `Owner language` should reset to the current behavior when the notification is sent in the language of the file owner
expected:
`php occ config:app:get core shareapi_public_notification_lang` produces an empty output in this case (no value set)

3. Switching the dropdown to any other value except `Owner language`  should update `shareapi_public_notification_lang` with a respective language code

## Screenshots (if appropriate):
notification lang

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by PVince81** [comment]
as discussed, ready for review again

File: lib/private/Helper/LocaleHelper.php [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+		$languages = [];
+
+		$availableCodes = $langFactory->findAvailableLanguages();
+		foreach ($availableCodes as $languageCode) {
+			$l = $langFactory->get('settings', $languageCode);
+
+			// TRANSLATORS this is a self-name of your language for the language switcher
+			$endonym = (string)$l->t('__language_name__');
+			//Check if the language name is in the translation file
+			// Fallback to hardcoded language name if translation is
```
Comment (phil-davis) [inline lib/private/Helper/LocaleHelper.php]: if translation is ???

Comment (VicDeo) [reply lib/private/Helper/LocaleHelper.php]: 'if it is not'

File: settings/Panels/Personal/Profile.php [hunk 2]
```diff
@@ -42,81 +43,59 @@ class Profile implements ISettings {
...
-			'en', 'es', 'fr', 'de', 'de_DE', 'ja', 'ar', 'ru', 'nl', 'it', 'pt_BR', 'pt_PT', 'da', 'fi_FI', 'nb_NO', 'sv', 'tr', 'zh_CN', 'ko'
-		];
-		$languageNames = $this->getLanguageCodes();
-		$languages= [];
-		$commonLanguages = [];
-		foreach ($languageCodes as $lang) {
-			$l = $this->lfactory->get('settings', $lang);
-			// TRANSLATORS this is the language name for the language switcher in the personal settings and should be the localized version
-			$potentialName = (string) $l->t('__language_name__');
-			if ($l->getLanguageCode() === $lang && \substr($potentialName, 0, 1) !== '_') {//first check if the language name is in the translation file
```
Comment (PVince81) [inline settings/Panels/Personal/Profile.php]: it looks like this complex logic got simplified in `getNormalizedLanguages` ?

please make sure you cover all these old code paths in a unit test in LocaleHelperTest

**Review by PVince81 — COMMENTED** [review]
Looks good so far.

I'm missing unit tests for the helper. This is important due to the complexity of the code you have simplified.

**Comment by PVince81** [comment]
Codecov didn't complain about missing tests for LocaleHelper because the code in question seems to be touched by the other tests. Would still be good to have a test class there to enforce (and document) the many different locale formats

**Comment by phil-davis** [comment]
It would be nice to have acceptance tests for the UI settings page. I made a general issue for that, an when @paurakhsharma gets initial tests, then it will be easy to add tests for this UI element.
(This is just a note FYI, not a blocker of this PR!)

File: settings/Panels/Personal/Profile.php (hunk 2 above)
Comment (VicDeo) [reply settings/Panels/Personal/Profile.php]: basically we check if the current lang has a self-name in a translation file, than fallback to harcoded language self-names and when everything else fails use language code as a self-name.

**Comment by VicDeo** [comment]
@PVince81 added some tests with explanations

File: lib/private/Helper/LocaleHelper.php [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+		'tr',
+		'zh_CN',
+		'ko'
+	];
+
+	/**
+	 * @var string[]
+	 */
+	private $languageCodes = [
+		'el' => 'Ελληνικά',
```
Comment (PVince81) [inline lib/private/Helper/LocaleHelper.php]: I wonder if we can get these from a PHP library in the future

**Review by PVince81 — APPROVED** [review]
:+1: looks good, thanks

**Comment by PVince81** [comment]
@VicDeo please backport

**Comment by mmattel** [comment]
I guess that this is documentation relevant at least becuase of:
``php occ config:app:get core shareapi_public_notification_lang``
In case, please open a doc issue and reference it

**Comment by mmattel** [comment]
This language definition for eMails could be useful for eMails sent when creating a new user!
@PVince81 @VicDeo

**Comment by VicDeo** [comment]
Stable10: https://github.com/owncloud/core/pull/32004

**Pull request merged**